geode-core/src/main/java/org/apache/geode/internal/cache/PartitionedRegion.java

```diff
@@ -1850,11 +1851,38 @@ public Lock getDistributedLock(Object key) throws IllegalStateException {
 
   /**
    * @since GemFire 5.0
+   *        Returns the statistics of the region by collecting the statistics of the
+   *        local buckets of the partitioned region.
    * @throws UnsupportedOperationException OVERRIDES
    */
   @Override
   public CacheStatistics getStatistics() {
-    throw new UnsupportedOperationException();
+    if (!statisticsEnabled) {
+      throw new StatisticsDisabledException(
+          String.format("Statistics disabled for region '%s'",
+              getFullPath()));
+    }
+    Set<BucketRegion> buckets = this.dataStore.getAllLocalBucketRegions();
+    long lastModifiedTime =
```

Comment:
This only computes the values when getStatistics is called. CacheStatistics is currently not a snapshot. You can call getStatistics once and then keep asking it for things like the hitCount or the lastModifiedTime and keep getting different results. I think you should preserve that behavior for PartitionedRegion. You can do this by overriding each of the CacheStatistics methods on PartitionedRegion to do the code you have here. So each time one of those methods is called it would recompute the value and return it. It is unclear to me why the atomics on AbstractRegion would need to be set (by calling this.setMissCount, etc.) since those atomics would never be used. This method could then go away since we can use the AbstractRegion.getStatistics

Reply:
Great. I will do it this way. The tests will also be modified so that we do not need to wait for 2 or 3 seconds in order for getStatistics to be invoked before each check for a count value.

Thanks a lot for the comments.

Reply:
You are also right that the set methods added to the AbstractRegion are not needed. I will remove them too.

Reply:
By the way, the JIRA ticket seemed to suggest that the implementation of the method should return "a snapshot of the state of the local buckets in the PR at the time getStatistics is called".
That's why I returned a snapshot in my initial implementation.
Should the phrasing of the ticket be updated?

Reply:
Okay, I updated the ticket. I know some of the other region ops on a partitioned region are a snapshot when they represent remote state (I think  Region.getEntry is like this). Maybe I was thinking of that when I filed that ticket. But since these stats only use local bucket state I think computing them each time is better than a snapshot.